addon/components/polaris-data-table.js

```diff
@@ -0,0 +1,534 @@
+
+    let { footerContent, truncate } = this.getProperties('footerContent', 'truncate');
+    let collapsed = this.get('table.scrollWidth') > this.get('dataTable.offsetWidth');
+
+    this.get('scrollContainer').scrollLeft = 0;
+    this.setProperties(assign(
+      {
+        collapsed,
+        heights: [],
+      },
+      this.calculateColumnVisibilityData(collapsed),
+    ));
+
+    scheduleOnce('afterRender', () => {
+      if (footerContent || !truncate) {
+        this.setHeightsAndScrollPosition();
+      }
+    });
+  },
+
+  scrollListener() {
+    if (this.get('isDestroying') || this.get('isDestroyed')) {
+      return;
+    }
+
+    this.setProperties(this.calculateColumnVisibilityData(this.get('collapsed')));
+  },
+
+  tallestCellHeights() {
+    let { footerContent, truncate, heights } = this.getProperties('footerContent', 'truncate', 'heights');
+    let rows = Array.from(this.get('table').getElementsByTagName('tr'));
+
+    if (!truncate) {
+      return (heights = rows.map((row) => {
+        let fixedCell = row.children[0];
+        return Math.max(row.clientHeight, fixedCell.clientHeight);
+      }));
+    }
+
+    if (footerContent) {
+      let footerCellHeight = rows[rows.length - 1].children[0].clientHeight;
+      heights = [footerCellHeight];
+    }
+
+    return heights;
+  },
+
+  addEventHandlers() {
+    this.get('eventHandlerParams').forEach(([ target, eventName, callbackName, ...rest ]) => {
+      target.addEventListener(eventName, this[callbackName].bind(this), ...rest);
+    });
+  },
+
+  removeEventHandlers() {
+    this.get('eventHandlerParams').forEach(([ target, eventName, callbackName, ...rest ]) => {
+      target.removeEventListener(eventName, this[callbackName].bind(this), ...rest);
+    });
+  },
+
+  /*
+   * Lifecycle hooks.
+   */
+  init() {
+    this._super(...arguments);
+
+    this.setProperties({
+      columnVisibilityData: [],
+      sorted: isPresent(this.get('sortable')),
+      heights: [],
+      preservedScrollPosition: {},
+    });
+  },
+
+  didInsertElement() {
+    this._super(...arguments);
+
+    this.handleResize();
+
+    this.addEventHandlers();
+  },
+
+  willDestroyElement() {
+    this.removeEventHandlers();
+  },
+
+  didUpdateAttrs() {
+    this._super(...arguments);
+
+    let truncate = this.get('truncate');
+    if (!truncate && this.get('previousTruncate')) {
+      this.handleResize();
+    }
+
+    this.set('previousTruncate', truncate);
+  },
+
+  actions: {
+    navigateTable(direction) {
+      let { scrollContainer, currentColumn, previousColumn } = this.getProperties('scrollContainer', 'currentColumn', 'previousColumn');
+
+      if (direction === 'right' && currentColumn) {
+        scrollContainer.scrollLeft = currentColumn.rightEdge;
+      } else if (previousColumn) {
+        scrollContainer.scrollLeft =
+          previousColumn.leftEdge < 10 ? 0 : previousColumn.leftEdge;
+      }
+
+      // TODO: use run loop instead of `requestAnimationFrame` here?
```

Comment:
Seems like `requestAnimationFrame` is just an overkill way of quickly calculating precise position values of columns? I don't know if swapping for run loop would accomplish whatever the intention of `requestAnimationFrame` here is, maybe just leave as-is unless you're certain that we need to throw this into one of the run loop cycles.

Reply:
My understanding is that `requestAnimationFrame` helps prevent jankiness on scrolling, resizing etc. by limiting the amount of work done to only that which can be fitted between two consecutive frame renders. With this in mind, using it here makes some kind of sense, although I'm slightly confused as to why it's not being used anywhere else (I'd expect the resize handler to be a candidate for its use, for example).

That being said, I'm not sure if using it directly here creates the possibility of interfering with the Ember run loop. The interesting-looking [`ember-run-raf`](https://github.com/runspired/ember-run-raf) addon seems to be advocating "using `run.scheduleOnce` + `requestAnimationFrame`" in combination, so maybe we should be doing that combo here, not sure...

Reply:
Oh wow I didn't realize there was an addon that tied the two together. If it's owned/authored by runspired then I trust it, but again for this purpose I don't think it's really an urgent need. It's up to you if you want to experiment with it but I think I'm fine to leave it as-is.

Reply:
Yeah agreed, that guy knows what he's doing (though I bet he wishes he was @vladucu 😉)! I wasn't intending to pull in the addon here though, just found it the other day and thought it was interesting.

Reply:
add a note here maybe to check `ember-run-raf` when testing this in the real world?

Reply:
From what I can see, `ember-run-raf` changes the _entire_ run loop to use `requestAnimationFrame` rather than adding the option to e.g. schedule on an animation frame queue, so probs not something we want to pull in to the host app 😬